src/OperationProcessor.ts

```diff
@@ -76,576 +18,145 @@ export interface OperationProcessor {
-
-        // In addition to removing the obsolete operation from the opHashToInfo map (after this if-block),
-        // If the obsolete operation was considered valid, then the parent's next link is invalid, need to remove the next link also.
-        // Else if the operation has a missing ancestor, then need to remove the operation from the list of waiting descendants of the missing ancestor.
-        if (opInfo.status === OperationStatus.Valid) {
-          this.invalidatePreviouslyValidOperation(opHash);
-        } else if (opInfo.status === OperationStatus.Unvalidated) {
-          const missingAncestor = opInfo.missingAncestor!;
-          const waitingDescendantsOfAncestor = this.waitingDescendants.get(missingAncestor)!;
-          waitingDescendantsOfAncestor.remove(opHash);
-        }
-
-        map.delete(opHash);
-      }
-    });
-
-    // ... deferred operations are in the deferred operations list
-    for (const [, opList] of this.deferredOperations.entries()) {
-      // The documentation for LinkedList does not say anything about iteration with list updates.
-      // To be safe, we store the deleted operations in a separate list and delete them after the
-      // iteration of opList.
-      const deletedOpList = new LinkedList<OperationHash>();
-      for (const opHash of opList) {
-        const operator = await this.operationStore.lookup(opHash);
-        if (operator.transactionNumber! >= transactionNumber) {
-          deletedOpList.append(opHash);
-        }
-      }
-
-      for (const deletedOpHash of deletedOpList) {
-        opList.remove(deletedOpHash);
-      }
-    }
+    return this.operationStore.delete(transactionNumber);
   }
 
   /**
    * Resolve the given DID to its DID Doducment.
    * @param did The DID to resolve. e.g. did:sidetree:abc123.
    * @returns DID Document of the given DID. Undefined if the DID is deleted or not found.
+   *
+   * Iterate over all operations in blockchain-time order extending the
+   * the operation chain while checking validity.
    */
   public async resolve (did: string): Promise<DidDocument | undefined> {
-    const didUniqueSuffix = did.substring(this.didMethodName.length);
-
-    await this.processDeferredOperationsOfDid(didUniqueSuffix);
+    let didDocument: DidDocument | undefined;
+    let previousOperation: Operation | undefined;
 
-    if (this.deletedDids.has(did)) {
-      return undefined;
-    }
-
-    const latestVersion = await this.last(didUniqueSuffix);
+    const didUniqueSuffix = did.substring(this.didMethodName.length);
 
-    // lastVersion === undefined implies we do not know about the did
-    if (latestVersion === undefined) {
-      return undefined;
+    const didOps = await this.operationStore.get(didUniqueSuffix);
+    for (const operation of didOps) {
+      if (await this.isValid(operation, previousOperation, didDocument)) {
+        didDocument = this.getUpdatedDocument(didDocument, operation);
+        previousOperation = operation;
```

Comment:
Consider adding a short-circuit check here:

// If this is a delete operation, this will be the last valid operation for this DID.
if (operation.type === OperationType.Delete) { break; }

Reply:
Good catch. Done.